File: ios/Runner/OnDeviceLlmHandler.swift
/// On-device LLM platform channel bridge.
///
/// iOS 26+ (device only): delegates to Apple Foundation Models.
/// Simulator / older iOS: returns available=false gracefully.
///
/// Channels:
///   MethodChannel  — ai.flutterclaw/on_device_llm
///   EventChannel   — ai.flutterclaw/on_device_llm_stream
import Flutter
import UIKit

// FoundationModels only exists on real devices with iOS 26+.
// The simulator has a stub TBD in the SDK but no runtime binary,
// so we must gate the import on !targetEnvironment(simulator).
#if !targetEnvironment(simulator) && canImport(FoundationModels)
import FoundationModels
#endif

class OnDeviceLlmHandler: NSObject, FlutterStreamHandler {

  private let methodChannel: FlutterMethodChannel
  private var streamTask: Task<Void, Never>?

  init(messenger: FlutterBinaryMessenger) {
    methodChannel = FlutterMethodChannel(
      name: "ai.flutterclaw/on_device_llm",
      binaryMessenger: messenger
    )
    super.init()
    methodChannel.setMethodCallHandler(handle(_:result:))
  }

  // MARK: - MethodChannel

  func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "check_availability":
      handleCheckAvailability(result: result)
    case "generate":
      guard let args = call.arguments as? [String: Any] else {
        result(FlutterError(code: "INVALID_ARG", message: "Expected Map args", details: nil))
        return
      }
      handleGenerate(args: args, result: result)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Availability

  private func handleCheckAvailability(result: @escaping FlutterResult) {
    #if targetEnvironment(simulator)
    result([
      "available": false,
      "platform": "ios",
      "reason": "Apple Foundation Models not available in iOS Simulator. Run on a real device.",
    ] as [String: Any])
    #elseif canImport(FoundationModels)
    if #available(iOS 26.0, *) {
      let model = SystemLanguageModel(useCase: .general)
      let langs = model.supportedLanguages.map { $0.maximalIdentifier }.joined(separator: ", ")
      switch model.availability {
      case .available:
        result([
          "available": true,
          "platform": "ios",
          "model_name": "apple-foundation",
          "context_window": 4096,
          "supports_streaming": true,
          "supports_tool_calling": false,
          "supported_languages": langs,
        ] as [String: Any])
      case .unavailable(let reason):
        let reasonStr: String
        switch reason {
        case .deviceNotEligible:
          reasonStr = "deviceNotEligible — device does not support Apple Intelligence (requires iPhone 15 Pro / iPhone 16+). Supported languages: \(langs)"
        case .appleIntelligenceNotEnabled:
          reasonStr = "appleIntelligenceNotEnabled — toggle is off or language not supported. Device language must match a supported language. Supported: \(langs)"
        case .modelNotReady:
          reasonStr = "modelNotReady — model still downloading. Check Settings → Apple Intelligence & Siri."
        @unknown default:
          reasonStr = "unavailable (unknown reason). Supported languages: \(langs)"
        }
        result([
          "available": false,
          "platform": "ios",
          "reason": reasonStr,
        ] as [String: Any])
      }
    } else {
      result([
        "available": false,
        "platform": "ios",
        "reason": "Requires iOS 26.0 or later.",
      ] as [String: Any])
    }
    #else
    result([
      "available": false,
      "platform": "ios",
      "reason": "FoundationModels framework not present in this build.",
    ] as [String: Any])
    #endif
  }

  // MARK: - Non-streaming generation

  private func handleGenerate(args: [String: Any], result: @escaping FlutterResult) {
    #if targetEnvironment(simulator)
    result(FlutterError(code: "UNAVAILABLE", message: "Not available in simulator.", details: nil))
    #elseif canImport(FoundationModels)
    guard #available(iOS 26.0, *) else {
      result(FlutterError(code: "UNAVAILABLE", message: "Requires iOS 26+.", details: nil))
      return
    }
    let messages  = args["messages"]  as? [[String: Any]] ?? []
    let maxTokens = args["max_tokens"] as? Int    ?? 1024
    let temperature = args["temperature"] as? Double ?? 0.7
    let prompt = buildPrompt(from: messages)

    Task {
      do {
        var options = GenerationOptions()
        options.maximumResponseTokens = maxTokens
        options.temperature = temperature
        let session  = LanguageModelSession()
        let response = try await session.respond(to: prompt, options: options)
        let content  = stripPromptEcho(response.content)
        let ptoks    = estimateTokens(prompt)
        let ctoks    = estimateTokens(content)
        DispatchQueue.main.async {
          result([
            "content": content,
            "finish_reason": "stop",
            "usage": ["prompt_tokens": ptoks, "completion_tokens": ctoks, "total_tokens": ptoks + ctoks],
          ] as [String: Any])
        }
      } catch {
        DispatchQueue.main.async {
          result(FlutterError(code: "GENERATION_FAILED", message: error.localizedDescription, details: nil))
        }
      }
    }
    #else
    result(FlutterError(code: "UNAVAILABLE", message: "FoundationModels not available.", details: nil))
    #endif
  }

  // MARK: - FlutterStreamHandler

  func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
    #if targetEnvironment(simulator)
    events(["type": "error", "message": "Not available in simulator."] as [String: Any])
    events(FlutterEndOfEventStream)
    #elseif canImport(FoundationModels)
    guard #available(iOS 26.0, *) else {
      events(FlutterError(code: "UNAVAILABLE", message: "Requires iOS 26+.", details: nil))
      events(FlutterEndOfEventStream)
      return nil
    }
    guard let args = arguments as? [String: Any] else {
      events(FlutterError(code: "INVALID_ARG", message: "Expected Map args.", details: nil))
      events(FlutterEndOfEventStream)
      return nil
    }
    let messages    = args["messages"]    as? [[String: Any]] ?? []
    let maxTokens   = args["max_tokens"]  as? Int    ?? 1024
    let temperature = args["temperature"] as? Double ?? 0.7
    let prompt      = buildPrompt(from: messages)
    let sink        = events

    streamTask = Task {
      do {
        var options = GenerationOptions()
        options.maximumResponseTokens = maxTokens
        options.temperature = temperature
        let session = LanguageModelSession()
        var previousContent = ""
        for try await snapshot in session.streamResponse(to: prompt, options: options) {
          let accumulated = snapshot.content
          let delta = String(accumulated.dropFirst(previousContent.count))
          previousContent = accumulated
          if !delta.isEmpty {
            DispatchQueue.main.async { sink(["type": "delta", "content": delta] as [String: Any]) }
          }
          if Task.isCancelled { break }
        }
        let ptoks = estimateTokens(prompt)
        let ctoks = estimateTokens(previousContent)
        DispatchQueue.main.async {
          sink(["type": "done", "finish_reason": "stop",
                "usage": ["prompt_tokens": ptoks, "completion_tokens": ctoks, "total_tokens": ptoks + ctoks]] as [String: Any])
          sink(FlutterEndOfEventStream)
        }
      } catch {
        DispatchQueue.main.async {
          sink(["type": "error", "message": error.localizedDescription] as [String: Any])
          sink(FlutterEndOfEventStream)
        }
      }
    }
    #else
    events(["type": "error", "message": "FoundationModels not available."] as [String: Any])
    events(FlutterEndOfEventStream)
    #endif
    return nil
  }

  func onCancel(withArguments arguments: Any?) -> FlutterError? {
    streamTask?.cancel()
    streamTask = nil
    return nil
  }

  // MARK: - Helpers

  /// Converts the messages array into a plain-text prompt.
  /// Uses a compact format optimised for small on-device models (~4K context).
  private func buildPrompt(from messages: [[String: Any]]) -> String {
    var parts: [String] = []
    for msg in messages {
      let role    = msg["role"]    as? String ?? "user"
      let content = msg["content"] as? String ?? ""
      guard !content.isEmpty else { continue }
      switch role {
      case "system":    parts.append("[System]\n\(content)")
      case "user":      parts.append("User: \(content)")
      case "assistant": parts.append("AI: \(content)")
      case "tool":      parts.append("Tool result: \(content)")
      default:          parts.append("\(role): \(content)")
      }
    }
    parts.append("AI:")
    return parts.joined(separator: "\n\n")
  }

  /// Strips echoed prompt content from the model's response.
  /// Some on-device models reproduce the conversation history instead of
  /// generating a new reply — especially when the context window is tight.
  private func stripPromptEcho(_ response: String) -> String {
    var cleaned = response.trimmingCharacters(in: .whitespacesAndNewlines)
    // Remove leading "AI:" prefix the model may echo.
    if cleaned.hasPrefix("AI:") {
      cleaned = String(cleaned.dropFirst(3)).trimmingCharacters(in: .whitespacesAndNewlines)
    }
    // Detect full echo: if the response contains multiple role markers
    // (User:/AI:/[System]) it is echoing the conversation.
    let hasMultipleRoles = cleaned.contains("User:") && cleaned.contains("AI:")
    let hasSystemEcho = cleaned.contains("[System]")
    if hasMultipleRoles || hasSystemEcho {
      if let range = cleaned.range(of: "AI:", options: .backwards) {
        cleaned = String(cleaned[range.upperBound...]).trimmingCharacters(in: .whitespacesAndNewlines)
      }
    }
    return cleaned
  }

  private func estimateTokens(_ text: String) -> Int { max(1, text.count / 4) }
}

File: android/app/src/main/kotlin/ai/flutterclaw/flutterclaw/OnDeviceLlmHandler.kt
package ai.flutterclaw.flutterclaw

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/// On-device LLM platform channel bridge — Android side.
///
/// Delegates to Gemini Nano via the Google AI Edge SDK (AICore).
/// Returns available=false gracefully on unsupported hardware so the Dart
/// layer can fall back to cloud providers.
///
/// Availability: Android 14+ (API 34) with compatible hardware
/// (Pixel 8+, Samsung Galaxy S24+, etc.). minSdk 26 is already satisfied.
///
/// Dependency (build.gradle.kts):
///   implementation("com.google.ai.edge.aicore:aicore:0.0.3-alpha01")
class OnDeviceLlmHandler(private val context: Context) :
    MethodChannel.MethodCallHandler, EventChannel.StreamHandler {

    private val mainHandler = Handler(Looper.getMainLooper())
    private val scope = CoroutineScope(Dispatchers.IO)
    private var streamJob: Job? = null

    // ── MethodChannel ────────────────────────────────────────────────────────

    fun handleMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "check_availability" -> handleCheckAvailability(result)
            "generate" -> handleGenerate(call, result)
            else -> result.notImplemented()
        }
    }

    // ── Availability ─────────────────────────────────────────────────────────

    private fun handleCheckAvailability(result: MethodChannel.Result) {
        scope.launch {
            val info = checkAvailabilityInternal()
            withContext(Dispatchers.Main) { result.success(info) }
        }
    }

    private fun checkAvailabilityInternal(): Map<String, Any> {
        return try {
            val model = createModel(maxOutputTokens = 256, temperature = 0.0f)
            model.close()
            mapOf(
                "available" to true,
                "platform" to "android",
                "model_name" to "gemini-nano",
                "context_window" to 4096,
                "supports_streaming" to true,
                "supports_tool_calling" to false,
            )
        } catch (e: Exception) {
            mapOf(
                "available" to false,
                "platform" to "android",
                "reason" to (e.message ?: "Gemini Nano not available on this device"),
            )
        }
    }

    // ── Non-streaming generation ──────────────────────────────────────────────

    private fun handleGenerate(call: MethodCall, result: MethodChannel.Result) {
        @Suppress("UNCHECKED_CAST")
        val messages = call.argument<List<Map<String, Any>>>("messages") ?: emptyList()
        val maxTokens = call.argument<Int>("max_tokens") ?: 1024
        val temperature = (call.argument<Double>("temperature") ?: 0.7).toFloat()
        val prompt = buildPrompt(messages)

        scope.launch {
            try {
                val model = createModel(maxOutputTokens = maxTokens, temperature = temperature)
                val response = model.generateContent(prompt)
                model.close()
                val content = stripPromptEcho(response.text ?: "")
                val promptTokens = estimateTokens(prompt)
                val completionTokens = estimateTokens(content)
                withContext(Dispatchers.Main) {
                    result.success(mapOf(
                        "content" to content,
                        "finish_reason" to "stop",
                        "usage" to mapOf(
                            "prompt_tokens" to promptTokens,
                            "completion_tokens" to completionTokens,
                            "total_tokens" to promptTokens + completionTokens,
                        ),
                    ))
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    result.error("GENERATION_FAILED", e.message ?: "Generation failed", null)
                }
            }
        }
    }

    // ── EventChannel (streaming) ──────────────────────────────────────────────

    override fun onListen(arguments: Any?, events: EventChannel.EventSink) {
        @Suppress("UNCHECKED_CAST")
        val args = arguments as? Map<String, Any> ?: emptyMap()
        @Suppress("UNCHECKED_CAST")
        val messages = args["messages"] as? List<Map<String, Any>> ?: emptyList()
        val maxTokens = (args["max_tokens"] as? Int) ?: 1024
        val temperature = ((args["temperature"] as? Double) ?: 0.7).toFloat()
        val prompt = buildPrompt(messages)

        streamJob = scope.launch {
            try {
                val model = createModel(maxOutputTokens = maxTokens, temperature = temperature)
                var totalContent = ""
                model.generateContentStream(prompt).collect { chunk ->
                    val delta = chunk.text ?: ""
                    totalContent += delta
                    withContext(Dispatchers.Main) {
                        events.success(mapOf("type" to "delta", "content" to delta))
                    }
                }
                model.close()
                val promptTokens = estimateTokens(prompt)
                val completionTokens = estimateTokens(totalContent)
                withContext(Dispatchers.Main) {
                    events.success(mapOf(
                        "type" to "done",
                        "finish_reason" to "stop",
                        "usage" to mapOf(
                            "prompt_tokens" to promptTokens,
                            "completion_tokens" to completionTokens,
                            "total_tokens" to promptTokens + completionTokens,
                        ),
                    ))
                    events.endOfStream()
                }
            } catch (e: Exception) {
                withContext(Dispatchers.Main) {
                    events.success(mapOf("type" to "error", "message" to (e.message ?: "Stream failed")))
                    events.endOfStream()
                }
            }
        }
    }

    override fun onCancel(arguments: Any?) {
        streamJob?.cancel()
        streamJob = null
    }

    // ── Helpers ───────────────────────────────────────────────────────────────

    /// Creates an AICore GenerativeModel instance.
    /// Throws if the model is not available on this device.
    private fun createModel(maxOutputTokens: Int, temperature: Float): com.google.ai.edge.aicore.GenerativeModel {
        val config = com.google.ai.edge.aicore.generationConfig {
            this.maxOutputTokens = maxOutputTokens
            this.temperature = temperature
            this.topK = 40
        }
        return com.google.ai.edge.aicore.GenerativeModel(
            generationConfig = config,
            downloadCondition = com.google.ai.edge.aicore.DownloadCondition.Builder()
                .setMinimumBatteryLevelPercent(0)
                .build(),
        )
    }

    /// Converts the messages array into a plain-text prompt.
    /// Uses a compact format optimised for small on-device models (~4K context).
    private fun buildPrompt(messages: List<Map<String, Any>>): String {
        val parts = mutableListOf<String>()
        for (msg in messages) {
            val role = msg["role"] as? String ?: "user"
            val content = msg["content"] as? String ?: ""
            if (content.isEmpty()) continue
            when (role) {
                "system" -> parts.add("[System]\n$content")
                "user" -> parts.add("User: $content")
                "assistant" -> parts.add("AI: $content")
                "tool" -> parts.add("Tool result: $content")
                else -> parts.add("$role: $content")
            }
        }
        parts.add("AI:")
        return parts.joinToString("\n\n")
    }

    /// Strips echoed prompt content from the model's response.
    /// Some on-device models reproduce the conversation history instead of
    /// generating a new reply — especially when the context window is tight.
    private fun stripPromptEcho(response: String): String {
        var cleaned = response.trim()
        // Remove leading "AI:" prefix the model may echo.
        if (cleaned.startsWith("AI:")) {
            cleaned = cleaned.removePrefix("AI:").trim()
        }
        // Detect full echo: if the response contains multiple role markers
        // (User:/AI:/[System]) it is echoing the conversation. Extract only
        // the text after the very last "AI:" marker.
        val hasMultipleRoles = cleaned.contains("User:") && cleaned.contains("AI:")
        val hasSystemEcho = cleaned.contains("[System]")
        if (hasMultipleRoles || hasSystemEcho) {
            val lastAI = cleaned.lastIndexOf("AI:")
            if (lastAI >= 0) {
                cleaned = cleaned.substring(lastAI + "AI:".length).trim()
            }
        }
        return cleaned
    }

    /// Rough token estimate: ~4 chars per token.
    private fun estimateTokens(text: String): Int = maxOf(1, text.length / 4)

    fun cleanup() {
        streamJob?.cancel()
        scope.cancel()
    }
}
